sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/SchemaAndRowProto.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp.bigquery;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.services.bigquery.model.TableRow;
+import com.google.cloud.bigquery.v3.RowOuterClass.Row;
+import com.google.cloud.bigquery.v3.RowOuterClass.StructField;
+import com.google.cloud.bigquery.v3.RowOuterClass.StructType;
+import com.google.cloud.bigquery.v3.RowOuterClass.StructValue;
+import com.google.cloud.bigquery.v3.RowOuterClass.TypeKind;
+import com.google.cloud.bigquery.v3.RowOuterClass.Value;
+import com.google.cloud.bigquery.v3.RowOuterClass.Value.ValueCase;
+import javax.annotation.Nullable;
+
+/**
+ * A wrapper for a {@link Row} and the {@link StructType} representing the schema of the
+ * table it was generated from. Also provides utility methods to extract data.
+ */
+public class SchemaAndRowProto {
+  private final StructType tableSchema;
+  private final Row row;
+  TableRow tableRow;
+
+  public SchemaAndRowProto(StructType tableSchema, Row row) {
+    this.tableSchema = checkNotNull(tableSchema, "tableSchema");
+    this.row = checkNotNull(row, "row");
+  }
+
+  public Row getRow() {
+    return row;
+  }
+
+  public StructType getSchema() {
+    return tableSchema;
+  }
+
+  /**
+   * Gets the value of the specified (possibly nested) field or null if it can't be found.
+   */
+  @Nullable
+  public Object get(String fieldName) {
```

Comment:
Ouch. Why is this needed? Is there a way to implement it so that code that calls .get() multiple times doesn't end up traversing and converting the row multiple times?

Reply:
This function does no conversion -- it simply traverses the underlying Row proto object and returns embedded values. Callers don't have to use it if they're comfortable parsing the Row object directly, but it's a common enough use case that it's worth providing a common implementation. Note this is significantly lower-overhead than the previous implementation which converted everything to a TableRow under the covers.